vault/identity_store_oidc_provider.go

```diff
@@ -0,0 +1,173 @@
+}
+
+const (
+	namedAssignmentPath = oidcTokensPrefix + "named_assignments/"
+)
+
+func oidcProviderPaths(i *IdentityStore) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: "oidc/assignment/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the assignment",
+				},
+				"entities": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity entity names",
+				},
+				"groups": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity group names",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.CreateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: i.pathOIDCReadAssignment,
+				},
+				logical.DeleteOperation: &framework.PathOperation{
+					Callback: i.pathOIDCDeleteAssignment,
+				},
+			},
+			ExistenceCheck:  i.pathOIDCAssignmentExistenceCheck,
+			HelpSynopsis:    "CRUD operations for OIDC assignments.",
+			HelpDescription: "Create, Read, Update, and Delete OIDC named assignments.",
+		},
+		{
+			Pattern: "oidc/assignment/?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: i.pathOIDCListAssignment,
+				},
+			},
+			HelpSynopsis:    "List OIDC assignments",
+			HelpDescription: "List all configured OIDC assignments in the identity backend.",
+		},
+	}
+}
+
+// pathOIDCCreateUpdateAssignment is used to create a new named assignment or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	name := d.Get("name").(string)
+
+	var assignment namedAssignment
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, namedAssignmentPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&assignment); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if entitiesRaw, ok := d.GetOk("entities"); ok {
+		assignment.Entities = entitiesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Entities = d.Get("entities").([]string)
+	}
+
+	if groupsRaw, ok := d.GetOk("groups"); ok {
+		assignment.Groups = groupsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Groups = d.Get("groups").([]string)
+	}
+
+	if err := i.oidcCache.Flush(ns); err != nil {
+		return nil, err
+	}
```

Comment:
I think we need to consider what we want to do for caching. We aren't using the cache for any reads, so I don't think there is a need to flush/invalidate `oidcCache` for this write. I suggest to remove use of `oidcCache` until we discuss a strategy for caching at the next dev sync.

Reply:
+1 on this. We likely have to think about caching when assignments need to be accessed during the validation process. The `IdentityStore` has `db *memdb.MemDB` available so we can probably use that for indexing.